File: geocoord/crs.py
"""Coordinate reference systems: the registry, and transformation to and from it.

Every system this application offers is described by one proj4 definition, and
both implementations run that same definition - pyproj here, proj4js in the
browser. That is deliberate and it is the only arrangement that keeps the two
halves in step.

Using ``CRS.from_epsg`` here instead would be more idiomatic and would silently
break the port. PROJ 9 keeps datum transformations in its own catalogue and
chooses among them *per point*, falling back to a ballpark offset outside an
operation's declared area; proj4js has no catalogue at all and applies whatever
``+towgs84`` term its definition carries, everywhere. Two points a kilometre
apart off Madeira can therefore be transformed by different operations on the
desktop and by the same one in the browser, and the results diverge by
hundreds of metres with nothing to indicate it. One fixed definition, used by
both, is deterministic.

The definitions live in ``crs_registry.json`` beside this file, generated from
the EPSG database and checked against it: every projected system agrees with
its authoritative EPSG transformation to zero metres over control points inside
its real coverage. The registry is data rather than code so that the browser
reads the identical bytes.

One trap worth naming, since it is invisible and expensive. ``+towgs84`` is read
with the position vector convention, while PROJ declares most of these
operations with the coordinate frame convention, and the two differ only in the
sign of the three rotations. With the signs the wrong way round Datum 73 lands
50 m out and Lisboa 61 m - errors small enough to look like ordinary survey
noise and large enough to matter. The registry generator negates them, and the
control points in the parity contract are what prove it stayed that way.
"""
from __future__ import annotations

import json
import pathlib
from functools import lru_cache

from pyproj import CRS, Transformer

_REGISTRY_PATH = pathlib.Path(__file__).with_name("crs_registry.json")

#: EPSG code (as a string) -> everything the application knows about it.
REGISTRY: dict = json.loads(_REGISTRY_PATH.read_text(encoding="utf-8"))

#: The system every internal coordinate is expressed in.
WGS84 = "4326"

#: proj4 definition of WGS84, the pivot every transformation passes through.
WGS84_PROJ4 = REGISTRY[WGS84]["proj4"]


def systems(kind: str | None = None) -> list:
    """The registry as a list, in registry order, optionally filtered by kind.

    ``kind`` is ``"geographic"`` or ``"projected"``; ``None`` returns both.
    """
    return [v for v in REGISTRY.values() if kind is None or v["kind"] == kind]


def get(code) -> dict:
    """One system's entry. Accepts the code as a string or an integer."""
    entry = REGISTRY.get(str(code))
    if entry is None:
        raise KeyError(f"unknown coordinate system: {code!r}")
    return entry


def utm_proj4(zone: int, south: bool = False, datum: str = "WGS84") -> str:
    """A proj4 definition for a UTM zone, on WGS84 or ETRS89.

    The generic escape hatch: zones 1 to 60, either hemisphere. It covers
    Angola (32S, 33S), Moçambique (36S, 37S), Cabo Verde, Guiné-Bissau and São
    Tomé e Príncipe the moment somebody needs them, without this application
    having to guess at national datums it cannot verify.
    """
    if not 1 <= int(zone) <= 60:
        raise ValueError(f"UTM zone must be between 1 and 60, got {zone!r}")
    ellipsoid = "+datum=WGS84" if datum.upper() == "WGS84" else "+ellps=GRS80"
    hemisphere = " +south" if south else ""
    return f"+proj=utm +zone={int(zone)}{hemisphere} {ellipsoid} +units=m +no_defs"


def utm_label(zone: int, south: bool = False) -> str:
    """The suffix used for a generic UTM zone's output columns, e.g. "UTM33S"."""
    return f"UTM{int(zone)}{'S' if south else 'N'}"


@lru_cache(maxsize=64)
def _transformer(source: str, target: str) -> Transformer:
    """Cached transformer between two proj4 definitions.

    Building one is expensive and the application builds the same handful over
    and over, once per column per rerun.
    """
    return Transformer.from_crs(CRS.from_proj4(source), CRS.from_proj4(target),
                                always_xy=True)


def transform(x, y, source: str, target: str):
    """Transform one coordinate between two proj4 definitions.

    Always x/y order - longitude then latitude for a geographic system, easting
    then northing for a projected one - regardless of what axis order the
    authority declares. Returns ``(None, None)`` for a value that cannot be
    transformed, which is what a point outside a projection's domain gives,
    rather than the infinities PROJ returns for one.
    """
    if x is None or y is None:
        return None, None
    try:
        out_x, out_y = _transformer(source, target).transform(float(x), float(y))
    except (ValueError, TypeError):
        return None, None
    if not (_finite(out_x) and _finite(out_y)):
        return None, None
    return out_x, out_y


def _finite(v) -> bool:
    return v is not None and v == v and abs(v) != float("inf")


def to_wgs84(x, y, source: str):
    """Transform into WGS84 longitude/latitude. Returns ``(lon, lat)``."""
    return transform(x, y, source, WGS84_PROJ4)


def from_wgs84(lon, lat, target: str):
    """Transform out of WGS84 into ``target``. Returns ``(x, y)``."""
    return transform(lon, lat, WGS84_PROJ4, target)


def esri_wkt(code=None, proj4: str | None = None) -> str:
    """ESRI WKT for a shapefile's ``.prj`` sidecar.

    Registry systems carry theirs precomputed, because only pyproj can produce
    it and the browser has to write the same bytes. For a generic UTM zone or a
    pasted definition there is nothing precomputed, so it is derived here - and
    the browser, which cannot, falls back to the WGS84 sidecar and says so.
    """
    if code is not None:
        return get(code)["esri_wkt"]
    if proj4 is None:
        raise ValueError("esri_wkt needs either a registry code or a proj4 string")
    return CRS.from_proj4(proj4).to_wkt("WKT1_ESRI")

File: web/src/core/crs.js
/**
 * Coordinate reference systems (JavaScript port).
 *
 * Mirrors `geocoord/crs.py`, and reads the *same file* it does: the registry is
 * imported from `geocoord/crs_registry.json`, not copied here. Two copies of
 * fourteen datum definitions would be two copies to keep in step, and the whole
 * reason the definitions are data rather than code is that neither side gets to
 * have its own.
 *
 * Both implementations run the same proj4 definition - proj4js here, pyproj
 * there. That is what keeps them in step, and it is not the obvious choice:
 * pyproj could look each system up by EPSG code instead. It must not. PROJ 9
 * keeps datum transformations in its own catalogue and picks among them *per
 * point*, falling back to a ballpark offset outside an operation's declared
 * area, while proj4js has no catalogue and applies whatever `+towgs84` its
 * definition carries, everywhere. Off Madeira that difference is hundreds of
 * metres, and nothing on either side would say so.
 *
 * proj4 is fetched on demand, like SheetJS and JSZip: a file already in WGS84
 * never needs it.
 */
import registry from '../../../geocoord/crs_registry.json' with { type: 'json' }

export const REGISTRY = registry

/** The system every internal coordinate is expressed in. */
export const WGS84 = '4326'

/** proj4 definition of WGS84, the pivot every transformation passes through. */
export const WGS84_PROJ4 = registry[WGS84].proj4

let proj4Promise = null
function loadProj4() {
  if (proj4Promise === null) {
    proj4Promise = import('proj4').then((m) => m.default ?? m)
  }
  return proj4Promise
}

/**
 * The registry as an array, in registry order, optionally filtered by kind.
 * `kind` is 'geographic' or 'projected'; omit it for both.
 */
export function systems(kind = null) {
  return Object.values(registry).filter((v) => kind === null || v.kind === kind)
}

/** One system's entry. Accepts the code as a string or a number. */
export function get(code) {
  const entry = registry[String(code)]
  if (entry === undefined) throw new Error(`unknown coordinate system: ${code}`)
  return entry
}

/**
 * A proj4 definition for a UTM zone, on WGS84 or ETRS89.
 *
 * The generic escape hatch: zones 1 to 60, either hemisphere. It covers Angola
 * (32S, 33S), Moçambique (36S, 37S), Cabo Verde, Guiné-Bissau and São Tomé e
 * Príncipe the moment somebody needs them, without this application having to
 * guess at national datums it cannot verify. Mirrors utm_proj4() in crs.py.
 */
export function utmProj4(zone, south = false, datum = 'WGS84') {
  const z = Number(zone)
  if (!Number.isInteger(z) || z < 1 || z > 60) {
    throw new Error(`UTM zone must be between 1 and 60, got ${zone}`)
  }
  const ellipsoid = datum.toUpperCase() === 'WGS84' ? '+datum=WGS84' : '+ellps=GRS80'
  return `+proj=utm +zone=${z}${south ? ' +south' : ''} ${ellipsoid} +units=m +no_defs`
}

/** The suffix used for a generic UTM zone's output columns, e.g. "UTM33S". */
export function utmLabel(zone, south = false) {
  return `UTM${Number(zone)}${south ? 'S' : 'N'}`
}

/**
 * Transform one coordinate between two proj4 definitions.
 *
 * Always x/y order - longitude then latitude for a geographic system, easting
 * then northing for a projected one - whatever axis order the authority
 * declares. Returns `[null, null]` for a value that cannot be transformed,
 * which is what a point outside a projection's domain gives, rather than the
 * infinities proj4 returns for one. Mirrors transform() in crs.py.
 */
export async function transform(x, y, source, target) {
  if (x === null || y === null || x === undefined || y === undefined) return [null, null]
  const proj4 = await loadProj4()
  let out
  try {
    out = proj4(source, target, [Number(x), Number(y)])
  } catch {
    return [null, null]
  }
  if (!Number.isFinite(out[0]) || !Number.isFinite(out[1])) return [null, null]
  return [out[0], out[1]]
}

/** Transform into WGS84 longitude/latitude. Returns `[lon, lat]`. */
export function toWgs84(x, y, source) {
  return transform(x, y, source, WGS84_PROJ4)
}

/** Transform out of WGS84 into `target`. Returns `[x, y]`. */
export function fromWgs84(lon, lat, target) {
  return transform(lon, lat, WGS84_PROJ4, target)
}

/**
 * Transform a whole column at once, loading proj4 only once.
 *
 * The per-value `transform` awaits the module every call, which is free after
 * the first but still a promise per row; a fifty-thousand-row file is worth
 * one await.
 */
export async function transformAll(pairs, source, target) {
  const proj4 = await loadProj4()
  const converter = proj4(source, target)
  return pairs.map(([x, y]) => {
    if (x === null || y === null || x === undefined || y === undefined) return [null, null]
    let out
    try {
      out = converter.forward([Number(x), Number(y)])
    } catch {
      return [null, null]
    }
    if (!Number.isFinite(out[0]) || !Number.isFinite(out[1])) return [null, null]
    return [out[0], out[1]]
  })
}

/**
 * ESRI WKT for a shapefile's `.prj` sidecar.
 *
 * Registry systems carry theirs precomputed by pyproj, which is the only side
 * that can produce it. A generic UTM zone or a pasted definition has none, so
 * the WGS84 sidecar is written and the caller is told - a `.prj` that lies
 * about its system is worse than one that is merely less specific.
 */
export function esriWkt(code = null) {
  if (code !== null && registry[String(code)] !== undefined) {
    return { wkt: registry[String(code)].esri_wkt, exact: true }
  }
  return { wkt: registry[WGS84].esri_wkt, exact: false }
}
